pkg/agent/controller/networkpolicy/status_controller.go

```diff
@@ -0,0 +1,230 @@
+// realized and all of its undesired rules have been removed.
+// For each new policy, SetRuleRealization is supposed to be called for each of its desired rules while
+// DeleteRuleRealization is supposed to be called for the removed rules.
+type StatusManager interface {
+	// SetRuleRealization updates the actual status for the given NetworkPolicy rule.
+	SetRuleRealization(ruleID string, policyID types.UID)
+	// DeleteRuleRealization deletes the actual status for the given NetworkPolicy rule.
+	DeleteRuleRealization(ruleID string)
+	// Resync triggers syncing status with the antrea-controller for the given NetworkPolicy.
+	Resync(policyID types.UID)
+	// Start the status sync loop.
+	Run(stopCh <-chan struct{})
+}
+
+// StatusController implements StatusManager.
+type StatusController struct {
+	nodeName string
+	// statusControlInterface knows how to update control plane NetworkPolicy status.
+	statusControlInterface networkPolicyStatusControlInterface
+	// ruleCache provides the desired state of NetworkPolicy rules.
+	ruleCache *ruleCache
+	// realizedRules keeps track of the realized NetworkPolicy rules.
+	realizedRules cache.Indexer
+	// queue maintains the UIDs of the NetworkPolicy that need to be processed.
+	queue workqueue.RateLimitingInterface
+}
+
+// realizedRule is the struct kept by StatusController for storing a realized rule.
+// It has policyID because "ruleCache" only keeps desired state of policies, so if a rule is no longer in a policy it
+// will be deleted immediately from "ruleCache" while we need to know these rules are actually uninstalled from
+// dataplane before their policies are considered realized.
+type realizedRule struct {
+	ruleID   string
+	policyID types.UID
+}
+
+func realizedRuleKeyFunc(obj interface{}) (string, error) {
+	return obj.(*realizedRule).ruleID, nil
+}
+
+func realizedRulePolicyIndexFunc(obj interface{}) ([]string, error) {
+	rule := obj.(*realizedRule)
+	return []string{string(rule.policyID)}, nil
+}
+
+func newStatusController(antreaClientProvider agent.AntreaClientProvider, nodeName string, ruleCache *ruleCache) *StatusController {
+	return &StatusController{
+		statusControlInterface: &networkPolicyStatusControl{antreaClientProvider: antreaClientProvider},
+		nodeName:               nodeName,
+		ruleCache:              ruleCache,
+		realizedRules: cache.NewIndexer(realizedRuleKeyFunc, cache.Indexers{
+			realizedRulePolicyIndex: realizedRulePolicyIndexFunc,
+		}),
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "networkpolicystatus"),
+	}
+}
+
+func (c *StatusController) SetRuleRealization(ruleID string, policyID types.UID) {
+	_, exists, _ := c.realizedRules.GetByKey(ruleID)
+	// This rule has been realized before. The current call must be triggered by group member updates, which doesn't
+	// affect the policy's realization status.
+	if exists {
```

Comment:
I guess you changed the design here? We no more track rule generation in agent?

Reply:
Yes, explained here https://github.com/vmware-tanzu/antrea/pull/1442#discussion_r519960601